### PR TITLE
Caching in makefile-tool

### LINKED ISSUE #110: Use caching in makefile example
From @dorranh (https://github.com/tweag/funflow2/pull/105#discussion_r650096792)

> I noticed that in the README we claim that this example caches build steps and initially thought that we had forgotten to port that logic over. After looking at the original code from funflow, I realized that we were not caching the build steps in the original example either sweat_smile. Since this wasn't implemented in the original example, I'm happy merging this PR and updating the caching in a later PR as well.
> 
> At first glance it looks like we might want to apply the caching wherever we call `compileFile` using `Control.Kernmantle.Caching.caching'` [link](https://tweag.github.io/funflow2/api/kernmantle-caching/html/Control-Kernmantle-Caching.html#v:caching-39-).

> Another example https://tweag.github.io/funflow2/tutorials/QuickReference.html#Caching-a-flow. I think we'll want to use caching' instead to avoid having to generate unique ids for each of the build steps manually.

### PR DESCRIPTION
**Pull request by vreuter**
Close #110 

@dorranh I ran up against ambiguity in the `core` type and inability to deduce a `ProvidesPosCaching` instance when I tried out `caching'`, so I went with regular `caching`, which seems to work. This compiled and when I execute repeatedly, I get the expected short circuit / early cutoff behavior that I don't with the version on `master`

### CODE REVIEW

File: funflow-examples/makefile-tool/src/Main.hs [hunk 1]
```diff
@@ -11,6 +11,7 @@ module Main where
 
 import Control.Arrow
 import Control.Exception (SomeException (..))
+import Control.Kernmantle.Caching (caching)
```
Comment (dorranh) [inline funflow-examples/makefile-tool/src/Main.hs:14]: While this works, we should probably use the re-exported `caching` for consistency with the tutorials.

```suggestion
import Funflow (caching)
```

File: .gitignore [hunk 2]
```diff
@@ -19,6 +19,7 @@ cabal.sandbox.config
 cabal.project.local
 
 .vscode
+.idea/
```
Comment (dorranh) [inline .gitignore:22]: No worries on this, just curious: is there a good IDEA extension for Haskell?

File: funflow-examples/makefile-tool/src/Main.hs [hunk 3]
```diff
@@ -105,12 +105,12 @@ buildTarget storeRoot mkfile target@(MakeRule targetNm deps cmd) = let
...
        let fullSrcFiles = Map.fromList $ zip neededSources contentSrcFiles
        -- Compile the target of the current rule.
-       compFile <- (compileFile storeRoot) -< (targetNm, fullSrcFiles, depFiles, cmd)
+       compFile <- caching targetNm (compileFile storeRoot) -< (targetNm, fullSrcFiles, depFiles, cmd)
        returnA -< compFile
 
 
 -- | Compiles a C file in a docker container.
-compileFile :: Path Abs Dir -> Flow (TargetFile, Map.Map SourceFile String, [Content File], Command) (Content File)
+compileFile :: Path Abs Dir -> Flow (TargetFile, Map.Map SourceFile String, [Path Abs File], Command) (Path Abs File)
```
Comment (dorranh) [inline funflow-examples/makefile-tool/src/Main.hs:113]: Just curious: what was the reason for needing to change the `Flow` input and output types with the introduction of `caching`? As far as I could tell caching should preserve the types of the inputs and outputs. 

https://github.com/tweag/kernmantle/blob/895c46820b7a44f3039de655b6efd73fd6544f39/caching/src/Control/Kernmantle/Caching.hs#L133

File: funflow-examples/package.yaml [hunk 4]
```diff
@@ -18,6 +18,7 @@ dependencies:
   - directory
   - funflow
   - kernmantle
+  - kernmantle-caching
```
Comment (dorranh) [inline funflow-examples/package.yaml:21]: We should be able to drop this dependency if we swap the import for `caching` above.

**Review by dorranh — COMMENTED** [review]
Nice, thanks for working on this! I didn't realize there was an easy way to get the build target names, and that seems like a nice way to structure the cache ids. Just a couple of general comments. Once they are addressed I'm happy merging.

File: .gitignore (hunk 2 above)
Comment (vreuter) [reply .gitignore:22]: Haha good eye! I didn't look yet, but yes this came up b/c it was the first time I'd opened the project in a JetBrains-developed editor. I ran into a merge conflict when I rebased before merging makefile-tool, and I'm just used to the interactive merge interface from PyCharm so went w/ that.

File: funflow-examples/makefile-tool/src/Main.hs (hunk 3 above)
Comment (vreuter) [reply funflow-examples/makefile-tool/src/Main.hs:113]: It was b/c of the `Store` constraint on the output type. I didn't see a hit for an `instance Store` in funflow or in kernmantle for `Content File`, but I did see one for `Path Abs File`.

Comment (dorranh) [reply funflow-examples/makefile-tool/src/Main.hs:113]: Gotcha that makes sense. Thanks for the additional details.

**Pull request merged**